### ISSUE TITLE
[bug] nil pointer when filter is not specified 

### ISSUE THREAD

**Issue by pxp928** [guacsec/guac#1107]
**Describe the bug**
By default, tests were run by specifying an empty filter for queries similar to below:

```
query HasSBOMQ1 {
  HasSBOM(hasSBOMSpec: {}) {
    ...allHasSBOMTree
  }
}
```

But it is possible to run the query without specifying the filter (not specifying `hasSBOMSpec: {}`). 
```
query HasSBOM {
  HasSBOM {
    uri
  }
```

When `hasSBOMSpec: {}` is not specified, it causes the `filter` itself to be `nil`

Also related to the issue: https://github.com/guacsec/guac/issues/1001 and https://github.com/guacsec/guac/issues/1105

Updated need to be made in the graphQL API filter for:

- artifact
- builder
- certifyBad
- certifyGood
- certifyScorecard
- certifyVex
- certifyVuln
- cve (will be deprecated)
- ghsa (will be deprecated)
- osv (will be deprecated)
- hashEqual
- hasSBOM
- hasSLSA
- hasSourceAt
- isDependency
- isOccurrence
- isVulnerability
- pkg
- pkEqual
- src

~~Fix and tests have to be added similar to the changes made in: https://github.com/guacsec/guac/pull/1106~~

EDIT: 

Based on the discussion below, updating the graphQL API to ensure that the filter is not null would ensure consistency across the backends and allow for other features (such as pagination) to be added in the future. 

For example:

https://github.com/guacsec/guac/blob/febfb5407256bc8eefde042777a1ed1805673106/pkg/assembler/graphql/schema/hasSBOM.graphql#L66

This should be changed to

```
  HasSBOM(hasSBOMSpec: HasSBOMSpec!): [HasSBOM!]!
```

This will still allow an empty filter but will guarantee that it is not null. Users will be prompted by the error message to add the filter.

**To Reproduce**
Run a query without the filter specified in the graphQL playground

**Expected behavior**
Return a graphQL error message that the filter is required for the query to function (even an empty filter)

**Screenshots**

**GUAC version**
main ([febfb54](https://github.com/guacsec/guac/commit/febfb5407256bc8eefde042777a1ed1805673106))

**Comment by mihaimaruseac** [comment]
We can also require the filter to be present (but allow empty)

**Comment by pxp928** [comment]
That is what we have currently, we would instead need to error if the filter is not provided. Seems like:

```
query HasSBOM {
  HasSBOM {
    uri
  }
```

is a valid query that many new users would type in causing an unneeded error. Better to handle the nil case IMO.

**Comment by mrizzi** [comment]
> That is what we have currently, we would instead need to error if the filter is not provided. Seems like:
> 
> ```
> query HasSBOM {
>   HasSBOM {
>     uri
>   }
> ```
> 
> is a valid query that many new users would type in causing an unneeded error. Better to handle the nil case IMO.

I think this is a very good point: new users would find it easier to be able to run "basic" queries (successfully) so handling `nil` seems a good way forward to me.

**Comment by mihaimaruseac** [comment]
https://github.com/guacsec/guac/blob/febfb5407256bc8eefde042777a1ed1805673106/pkg/assembler/graphql/schema/hasSBOM.graphql#L66

This should be

```
  HasSBOM(hasSBOMSpec: HasSBOMSpec!): [HasSBOM!]!
```

Will still allow empty filter, but will guarantee that it is not null. Users will be prompted by the error message to add the filter and this will be consistent with pagination support, etc. that will be implemented

**Comment by pxp928** [comment]
Ah if it's needed for support for pagination and other future features, it makes sense to do ensure the filter is set.

**Comment by mihaimaruseac** [comment]
I mean, we could also make pagination, etc. be optional and support the query with no args, but then we have to take this into account for all backends instead of fixing it in just one place.

No strong feelings either way.

**Comment by pxp928** [comment]
Better to fix it via the API (in one place) instead of each backend to avoid confusion and make things optional. I will update the above to reflect this

**Comment by mrizzi** [comment]
I've given it a try following @mihaimaruseac's suggestion and this is how it looks from GraphQL playground and the response which looks quite clear for the user in terms of understanding what is missing.

![Screenshot from 2023-07-31 19-10-03](https://github.com/guacsec/guac/assets/7288588/6c956993-841b-4102-8030-04fb251feb7d)

**Comment by mrizzi** [comment]
@mihaimaruseac for the pagination, if mandatory, maybe it would be good to evaluate having default values so that if not provided, then everything will work anyway

**Comment by pxp928** [comment]
> some spec-less queries - like the HasSBOM above - won't be valid any further)

@mrizzi not sure I understand. Every noun/verb can have a spec/filter applied to it. See here for examples: https://github.com/guacsec/guac/tree/main/pkg/assembler/graphql/examples

So this would not be a breaking change just mandate a filter be set (even if empty)

**Comment by mrizzi** [comment]
Well, it might be out of scope for this issue but I'm very interested in the versioning scheme because, as @mihaimaruseac pointed out, if some changes have to be implemented in all the backends (e.g. pagination), it could be that not all the backends will be ready at the same time so having a way to state which backend supports which GraphQL scheme version would be awesome.

**Comment by mrizzi** [comment]
Could you guys assign this issue to me please?
I tried locally to apply the changes in order for me to better understand guac and how it works/builds.

**Comment by pxp928** [comment]
> Well, it might be out of scope for this issue but I'm very interested in the versioning scheme because, as @mihaimaruseac pointed out, if some changes have to be implemented in all the backends (e.g. pagination), it could be that not all the backends will be ready at the same time so having a way to state which backend supports which GraphQL scheme version would be awesome.

For the time being, we can assume that there will be many breaking changes to the API. As the project progresses, we are looking to version the schema for various integration reasons.

**Comment by mrizzi** [comment]
> > Well, it might be out of scope for this issue but I'm very interested in the versioning scheme because, as @mihaimaruseac pointed out, if some changes have to be implemented in all the backends (e.g. pagination), it could be that not all the backends will be ready at the same time so having a way to state which backend supports which GraphQL scheme version would be awesome.
> 
> For the time being, we can assume that there will be many breaking changes to the API. As the project progresses, we are looking to version the schema for various integration reasons.

Totally makes sense, thanks.

I was wondering if and how the versioning could also help towards backends implementations.

I mean, let's assume a new GraphQL file schema is defined for the new schemas version (eg `1.1.0`)
This implies all the backends have to implement the methods for managing the newly added schema.
But the implementations won't take all the same time so it could be that backend X is "Graphql schemas v1.1.0 compatible" because it already implemented the new schema and, at the same time, backend Y is "Graphql schemas v1.0.0 compatible" because it has no implementation for the v1.1.0 changes yet.
Could this be a valid approach for "decoupling" the GraphQL schema changes and the multiple backends implementations?

**Comment by mrizzi** [comment]
I see what you mean, thanks :+1: 
From timeline perspective it fully makes sense.

And what do you think about the guac versions perspective?
I mean, could be a guac version supporting multiple graphql schema versions for different backends?
E.g. Guac v1.2.3 supports

- graphql schema v.1.1.0 for backends X and Y
- graphql schema v.1.0.0 for backends W and Z

In order to not have to wait for backends W and Z to implement graphql schema v.1.1.0 before doing the Guac v1.2.3 release.
From the user perspective, even if the backends are transparent, once one has been deployed the user should keep using that one (I haven't found anything about backends changes with data migration) and in this way it would be clear which graphql schema will be available upgrading to Guac v1.2.3

WDYT?
